service/consul/datacenterWatch.go

```diff
@@ -1,109 +1,229 @@
-			keys := make(map[string]bool)
-			instancersToAdd := make(service.Instancers)
-
+		case <-ticker.C:
 			datacenters, err := getDatacenters(logger, client, options)
 
 			if err != nil {
 				continue
 			}
 
-			for _, w := range options.watches() {
-				if w.CrossDatacenter {
-					for _, datacenter := range datacenters {
-						w.QueryOptions.Datacenter = datacenter
-
-						// create keys for all datacenters + watched services
-						key := newInstancerKey(w)
-						keys[key] = true
-
-						// don't create new instancer if it is already saved in environment's instancers
-						if currentInstancers.Has(key) {
-							continue
-						}
-
-						// don't create new instancer if it was already created and added to the new instancers map
-						if instancersToAdd.Has(key) {
-							logger.Log(level.Key(), level.WarnValue(), logging.MessageKey(), "skipping duplicate watch", "service", w.Service, "tags", w.Tags, "passingOnly", w.PassingOnly, "datacenter", w.QueryOptions.Datacenter)
-							continue
-						}
-
-						// create new instancer and add it to the map of instancers to add
-						instancersToAdd.Set(key, newInstancer(logger, client, w))
-					}
+			d.UpdateInstancers(datacenters)
+
+		}
+
+	}
+}
+
+func (d *DatacenterWatcher) UpdateInstancers(datacenters []string) {
+	keys := make(map[string]bool)
+	instancersToAdd := make(service.Instancers)
+
+	options := d.options
+	environment := d.environment
+	logger := d.logger
+	currentInstancers := environment.Instancers()
+
+	for _, w := range options.watches() {
+		if w.CrossDatacenter {
+			for _, datacenter := range datacenters {
+
+				//check if datacenter is part of inactive datacenters list
+				d.lock.RLock()
+				_, found := d.inactiveDatacenters[datacenter]
+				d.lock.RUnlock()
+
+				if found {
+					logger.Log(level.Key(), level.InfoValue(), logging.MessageKey(), "datacenter set as inactive", "datacenter name: ", datacenter)
+					continue
 				}
+				w.QueryOptions.Datacenter = datacenter
+
+				// create keys for all datacenters + watched services
+				key := newInstancerKey(w)
+				keys[key] = true
+
+				// don't create new instancer if it is already saved in environment's instancers
+				if currentInstancers.Has(key) {
+					continue
+				}
+
+				// don't create new instancer if it was already created and added to the new instancers map
+				if instancersToAdd.Has(key) {
+					logger.Log(level.Key(), level.WarnValue(), logging.MessageKey(), "skipping duplicate watch", "service", w.Service, "tags", w.Tags, "passingOnly", w.PassingOnly, "datacenter", w.QueryOptions.Datacenter)
+					continue
+				}
+
+				// create new instancer and add it to the map of instancers to add
+				instancersToAdd.Set(key, newInstancer(logger, d.environment.Client(), w))
 			}
+		}
+	}
+
+	environment.UpdateInstancers(keys, instancersToAdd)
+
+}
 
-			environment.UpdateInstancers(keys, instancersToAdd)
+func (d *DatacenterWatcher) DatacentersListener() chrysom.ListenerFunc {
```

Comment:
```suggestion
func (d *datacenterWatcher) datacentersListener() chrysom.ListenerFunc {
```
This can be private and set directly to the chrysom config